Compiler suggests unnecessary try into

I updated a dependency and the type of a struct field changed from an `i32` to a `u16`.

I am assigning a `u16` value to this struct field. Before, I was calling `i32::from` to convert the `u16` to an `i32`. With the upgrade this is unnecessary since the field  and the value are both `u16`.

The rust compiler suggests this fix:

```
error[E0308]: mismatched types
  --> artichoke-frontend/src/parser.rs:90:37
   |
90 |             (*self.parser).lineno = i32::from((*self.context).lineno);
   |                                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected u16, found i32
help: you can convert an `i32` to `u16` and panic if the converted value wouldn't fit
   |
90 |             (*self.parser).lineno = i32::from((*self.context).lineno).try_into().unwrap();
   |                                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: aborting due to previous error

For more information about this error, try `rustc --explain E0308`.
error: Could not compile `artichoke-frontend`.
warning: build failed, waiting for other jobs to finish...
error: build failed
```

But it should really suggest removing the `i32::from` and using the `u16` directly.

<!-- TRIAGEBOT_START -->

<!-- TRIAGEBOT_ASSIGN_START -->

<!-- TRIAGEBOT_ASSIGN_DATA_START$${"user":null}$$TRIAGEBOT_ASSIGN_DATA_END -->

<!-- TRIAGEBOT_ASSIGN_END -->
<!-- TRIAGEBOT_END -->

Similarly, this code should suggest replacing `i32::from` with `u16::from`.

```rust
#[derive(Debug)]
struct Member(u16);

pub fn main() {
    let m = Member(i32::from(0_u8));
    println!("{:?}", m);
}
```

```
   Compiling playground v0.0.1 (/playground)
error[E0308]: mismatched types
 --> src/main.rs:5:20
  |
5 |     let m = Member(i32::from(0_u8));
  |                    ^^^^^^^^^^^^^^^ expected u16, found i32
help: you can convert an `i32` to `u16` and panic if the converted value wouldn't fit
  |
5 |     let m = Member(i32::from(0_u8).try_into().unwrap());
  |                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: aborting due to previous error

For more information about this error, try `rustc --explain E0308`.
```

[playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=4cd158b1081140e5acc54a001b2a6392)

--- a/src/librustc_typeck/check/demand.rs
+++ b/src/librustc_typeck/check/demand.rs
@@ -587,6 +587,30 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                 return false;
             }
         }
+        if let hir::ExprKind::Call(path, args) = &expr.node {
+            if let (
+                hir::ExprKind::Path(hir::QPath::TypeRelative(base_ty, path_segment)),
+                1,
+            ) = (&path.node, args.len()) {
+                // `expr` is a conversion like `u32::from(val)`, do not suggest anything (#63697).
+                if let (
+                    hir::TyKind::Path(hir::QPath::Resolved(None, base_ty_path)),
+                    sym::from,
+                ) = (&base_ty.node, path_segment.ident.name) {
+                    if let Some(ident) = &base_ty_path.segments.iter().map(|s| s.ident).next() {
+                        match ident.name {
+                            sym::i128 | sym::i64 | sym::i32 | sym::i16 | sym::i8 |
+                            sym::u128 | sym::u64 | sym::u32 | sym::u16 | sym::u8 |
+                            sym::isize | sym::usize
+                            if base_ty_path.segments.len() == 1 => {
+                                return false;
+                            }
+                            _ => {}
+                        }
+                    }
+                }
+            }
+        }
 
         let msg = format!("you can convert an `{}` to `{}`", checked_ty, expected_ty);
         let cast_msg = format!("you can cast an `{} to `{}`", checked_ty, expected_ty);

--- a/src/test/ui/suggestions/mismatched-types-numeric-from.rs
+++ b/src/test/ui/suggestions/mismatched-types-numeric-from.rs
@@ -1,0 +1,3 @@
+fn main() {
+    let _: u32 = i32::from(0_u8); //~ ERROR mismatched types
+}

--- a/src/test/ui/suggestions/mismatched-types-numeric-from.stderr
+++ b/src/test/ui/suggestions/mismatched-types-numeric-from.stderr
@@ -1,0 +1,9 @@
+error[E0308]: mismatched types
+  --> $DIR/mismatched-types-numeric-from.rs:2:18
+   |
+LL |     let _: u32 = i32::from(0_u8);
+   |                  ^^^^^^^^^^^^^^^ expected u32, found i32
+
+error: aborting due to previous error
+
+For more information about this error, try `rustc --explain E0308`.

